### ISSUE TITLE
[Suggestion] Create some tutorials on how to use py-videocore6

### ISSUE THREAD

**Issue by filiphhh** [Idein/py-videocore6#43]
Hi!

I have been following py-videocore6 since you first made it public and py-videocore before that.
I'd like to thank you for simplifying QPU-programming on the Raspberry Pi's.
I have been trying to better learn how one would go about and better utilize the QPUs but I think it's difficult to find resources for it. I would love to see some tutorials on how to write applications with py-videocore6 and on how to parallellize the programs to fully utilize videocores potential.

Thank you @Terminus-IMRC, @notogawa and Idein for all your hard work!

**Comment by Terminus-IMRC** [comment]
Thank you for supporting us!
We also think we need some tutorials for beginners, but currently, no tutorials exist except for the Japanese one: https://qiita.com/9_ties/items/15ab7fa198991a61a3a9

Because the instruction set of VC6 QPU is very similar to the one of VC4 QPU, you can learn how QPU basically works (add/mul ALU dual-issue, three branch delay slots, TMU unit, etc.) from the [VideoCore IV 3D Architecture Reference Guide](https://docs.broadcom.com/docs/12358545).

Though there is no publicly available documentation for VC6 QPU, you can gain an understanding of it from working examples.
I added some simple example programs to this repository, which may help you when you write VC6 QPU codes:

- [`examples/summation.py`](https://github.com/Idein/py-videocore6/blob/master/examples/summation.py) adds up 32-bit integers in an array.
With this QPU code, you can learn how to fill the eight-depth TMU read request queue to hide cache/memory latency.
- [`examples/memset.py`](https://github.com/Idein/py-videocore6/blob/master/examples/memset.py) sets a single 32-bit value to an array.
You will find out that there is no request queue for TMU write.
- [`examples/scopy.py`](https://github.com/Idein/py-videocore6/blob/master/examples/scopy.py) copies an array, which combines the above two facts.

These codes support multiple-QPU execution up to eight, where you can see how to assign input/output memory area to each QPU.

Also, notogawa added matrix-matrix multiplication code [`examples/sgemm.py`](https://github.com/Idein/py-videocore6/blob/master/examples/sgemm.py).
The innermost loop of this code utilizes in-QPU vector rotation to reduce the number of memory loads/stores.

In conclusion, I recommend you to start writing a primitive program (simple memory read/write or array addition/subtraction/multiplication) by referring to the examples.
Then you may consider how to achieve the theoretical 32 [Gflop/s] peak performance (by utilizing the register files and TMU/L2 caches).

**Comment by Terminus-IMRC** [comment]
We've just released other VC6 QPU examples: https://github.com/Idein/qmkl6

**Comment by filiphhh** [comment]
Thank you for your quick replies!
I would never have found the Japanese tutorial if you didn't share the link to it, Google Translate seems to do a pretty good job at translating it.
Great to see QMKL for rpi4 too!

I hope that your libraries will get some more traction in the community as it unlocks a lot more power in these little devices.
Thank you for all the low level rpi resources you have produced!

**Comment by Abhiram001-coder** [comment]
Hi there, first of all  I would like to thank @Terminus-IMRC @notogawa for developing py-videocore6 driver, I have tried py-videocore6, it worked for me, now i am trying to use it as a application driver, to implement the April Tag detection using pi4 VC6, I am facing issues while mapping of these two lines from "videocore6.driver import Driver
from videocore6.assembler import qpu" and I am getting no module named vediocore VI, in a single folder py-videocore VI is in one driver folder and in another folder I am having my april tag detection tag, and I am finding it difficult to mapping these two, hoping for valuable suggestions from the community.

**Comment by notogawa** [comment]
Is [`ioctl-opt`](https://pypi.org/project/ioctl-opt/) correctly installed and available on your [`PYTHONPATH`](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH)?

**Comment by Abhiram001-coder** [comment]
Hi @notogawa as I am trying to implement april tag, it uses TAPI based Opencv2 commands which calls OpenCL commands which maps with v3d GPU hardware registers, but in the case of RPI4 Model B ver. 1.5 I came to know that the developed CL libraries are not compatible i guess, referred from, https://github.com/SatDump/SatDump/issues/877 so is it better to go ahead with CLVK approach, kindly suggest me

**Comment by notogawa** [comment]
@Abhiram001-coder 

In https://github.com/Idein/py-videocore6/issues/43#issuecomment-3483082824
> If possible, try reproducing it in a newly set up environment on your side.

means "try reinstall OS from zero (using another SD card)". not only do

> I have followed the below mentioned steps to get that error




About https://github.com/Idein/py-videocore6/issues/43#issuecomment-3484198545, I'm not in a position to advise on your technical choices, and this topic is no longer related to the py-videocore6 library.


That's help for you (and other user). You should read [this document](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax). Your comment’s formatting is broken. It is hard to read your info. And then, please check your comment (and its format) "Preview" before submit.
